### DISCUSSION TITLE
[SRP Client] How to create an SRP service subtype using OT methods

### DISCUSSION THREAD

**Discussion by estelleOl** [openthread/openthread#8183] [Q&A] [answered]
I try to create an SRP service subtype without using cli commands.
In CLI command it would be :  "srp client service add ins1 _test._udp,_sub1 111 1 1" 

So in my code I have created a service to add and tried to had it, but I do not succeed to have the right service name with sub type.

My code to add a basic type service : 
static otSrpClientService     services[2]              = {0};

services[0].mInstanceName = "ins1";
services[0].mName = "_test._udp";
services[0].mPort = 111;
services[0].mPriority = 1;
services[0].mWeight = 1;

otSrpClientAddService(NULL, &services[0]);

I have tried to modify the service name to declared a sub type but it did not work : 
static otSrpClientService     services[2]              = {0};

services[0].mInstanceName = "ins1";
services[0].mName = "_test._udp_sub1._sub";
services[0].mPort = 111;
services[0].mPriority = 1;
services[0].mWeight = 1;

otSrpClientAddService(NULL, &services[0]);

Can someone explain me what I did not understood in the service subtype creation ?

Thanks

**Comment by jwhui** [comment, marked as answer]
You need to add the sub-type labels separately via the [`mSubTypeLabels`](https://openthread.io/reference/struct/ot-srp-client-service#msubtypelabels) array in the [`otSrpClientService`](https://openthread.io/reference/struct/ot-srp-client-service#summary) struct.

You can also see how the CLI implementation processes sub-types:
https://github.com/openthread/openthread/blob/0756d1ea8fee17eba0cdee42a34ea353413a4f99/src/cli/cli_srp_client.cpp#L386-L406

**Reply by estelleOl** [reply]
Thank you for your reponse, I cannot activate  OPENTHREAD_CONFIG_SRP_CLIENT_BUFFERS_ENABLE so I cannot use the otSrpClientBuffersGetSubTypeLabelsArray ... . 

Can I directly modify the mSubTypeLabels ?

Something like that ?
services[0].mSubTypeLabels[0]="test";

**Reply by jwhui** [reply]
No, the caller needs to provide the buffer for the `char*` array itself.

**Reply by estelleOl** [reply]
Thank you for your help !

**Reply by abtink** [reply]
Let me also highlight couple of notes (which I hope helps):
-  The `mSubTypeLabels` array (of strings or `char *`) MUST end with `NULL` (which indicate end of the array).
-  The entries in `otSrpClientService` must persist and stay same after an instance of this struct is passed to OpenThread.

https://github.com/openthread/openthread/blob/2cab708443ece91555a9ba6d238b5908506ac409/include/openthread/srp_client.h#L84-L102
